quinn/src/platform/unix.rs

```diff
@@ -332,19 +336,55 @@ pub fn caps() -> UdpCapabilities {
 
 const CMSG_LEN: usize = 80;
 
+#[inline]
+fn in_addr(addr: &Ipv4Addr) -> libc::in_addr {
+    libc::in_addr { s_addr: u32::from_ne_bytes(addr.octets()) }
+}
+
+enum CSocketAddr {
```

Comment:
Why not use `socket2` instead of reinventing this, if that already provides the guarantee?

And I think just changing the socket type in `Transmit` to `socket2::SocketAddr` might still be an interesting option. The main advantage that std could yield is no-std compatibility. But a quic library only really matters for platforms which have a good network stack - and this case it is super likely that `socket2` is available for them too.

Reply:
> Why not use socket2 instead of reinventing this

Interesting idea. Apparently it's already used by dependencies of quinn so adding it as dependency wouldn't bloat the cargo graph. However, I can't find a way to get a mutable pointer without violating const correctness. It can only obtain const pointers.

> And I think just changing the socket type in Transmit to socket2::SocketAddr might still be an interesting option.

I've thought about that but it would have involved changes in all the places the Transmit struct is used. I also wanted to make a patch that one can backport safely to older releases without breaking the API.

Reply:
The macos issues found by CI in particular seem like a good argument to not duplicate this functionality, assuming socket2 is small enough that we shouldn't be looking to get rid of it in the first place. I think the mutability issue can be avoided by using [`from_raw_parts`](https://docs.rs/socket2/0.3.19/socket2/struct.SockAddr.html#method.from_raw_parts). A minimum-impact, backportable change does seem like a good place to start.

Reply:
> I think the mutability issue can be avoided by using from_raw_parts.

That function takes in a const pointer of the C representation and can be used to obtain the Rust representation. What we need is the opposite way: given the Rust representation, how can you obtain a *mutable* pointer to the C representation?

I don't think there is a fix for this without changes of the socket2 crate.

Reply:
> given the Rust representation, how can you obtain a mutable pointer to the C representation?

Why do you need this?

Reply:
@Ralith check the code. We do `hdr.msg_name = name;`. That hdr is of type [`libc::msghdr`](https://docs.rs/libc/0.2.82/libc/struct.msghdr.html) and its `msg_name` member is of type `*mut c_void`. So `name` has to be a mutable thing for const correctness.

See also https://github.com/rust-lang/rust-clippy/issues/4774#issuecomment-565651216 ... the `*mut` may have been a `*const` previously, but it may *not* originate from a `&T` reference but has to come from a `&mut T`. socket2's `as_ptr` function though takes in a `&self` and not a `&mut self`.

Reply:
I guess what one *could* do would be to copy from the `*const` pointer to a newly assigned location and then create a mutable pointer for that. It'd be a bit ugly but the code would only live in one location. I can change it if you want.

Reply:
Hmm, is this technically a bug with `libc` then, or would it need two different versions of `msghdr` (a mutable and non mutable one)?

the iov which is passed is also am mutable pointer, but we are passing it an immutable one. IDK, we know that libc doesn't mutate those things and just requires requires an immutable pointer for sending.

Reply:
Casting `*const` to `*mut` is perfectly fine here because we know no mutation will occur. Pointers are not references.

Reply:
> we know that libc doesn't mutate those things

Do you have sources for this claim? Or in other words, does the posix spec mandate this or can a random libc implementation just mutate if it wants to?

Reply:
Hmmm....

> The msghdr structure is used to minimize the number of directly supplied parameters to the recvmsg() and sendmsg() functions. This structure is used as a value- result parameter in the recvmsg() function and value only for the sendmsg() function.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_socket.h.html

I guess I'll just use socket2 then and just cast away the constness.